### PR TITLE
switch to the smart-arg package

### PR DESCRIPTION
**Pull request by jakiejj**
# Description
The docs of smart-arg are temporarily hosted [here](http://jjia-ld2:1080/) and should be available at readthedocs soon.

Change `Args` -> `DetextArg` to make more sense.

Refactored the detext logic to `run_detext` method which takes a single argument class of `DetextArg`, so that it makes more sense.

Moved the conversion from `DetextArg` to `HParams` as late as possible to utilize more IDE helps (type hints, auto-completion, and other error detection)

Moved `DetextArg` manipulation logic to where it's supposed to be (in the argument class for proper encapsulation).

## List all changes 
Please list all changes in the commit.
* `None` should only be assigned to `Optional` types
* make more use of typed parameter
* demo to use typed argument in test

## Type of changes
- [x] New feature (non-breaking change which adds functionality)
Some of the changes are not strictly non-breaking but should be at fairly low risk.

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/detext/run_detext.py [hunk 1]
```diff
@@ -132,100 +136,117 @@ class Args(NamedTuple):
...
+
+def get_hparams(argument: DetextArg = None):
     """
     Get hyper-parameters.
+    To be removed.
     """
-    hparams = tf.contrib.training.HParams(**Args()._asdict())
+    import warnings
+    warnings.warn("get_hparams is deprecated, call `run_detext` directly")
+    hparams = tf.contrib.training.HParams(**(argument or DetextArg.__from_argv__(error_on_unknown=False))._asdict())
```
Comment (StarWang) [inline src/detext/run_detext.py:191]: As `Hparams` is directly imported, can we replace this without the tf.xxx package prefix?

File: src/detext/run_detext.py [hunk 2]
```diff
@@ -132,100 +136,117 @@ class Args(NamedTuple):
...
-    task_ids: List[int] = None  # All types of task IDs for multitask training. E.g., 1,2,3
+    task_ids: Optional[List[int]] = None  # All types of task IDs for multitask training. E.g., 1,2,3
     _task_ids = __comma_split_list(int)
-    task_weights: List[float] = None  # Weights for each task specified in task_ids. E.g., 0.5,0.3,0.2
+    task_weights: Optional[List[float]] = None  # Weights for each task specified in task_ids. E.g., 0.5,0.3,0.2
     _task_weights = __comma_split_list(float)
 
-
-def get_hparams():
+    def __late_init__(self):
```
Comment (StarWang) [inline src/detext/run_detext.py:156]: This seems to be specific to smart-args. Can you add some comments to the function wrt what this is doing?

File: src/detext/run_detext.py (hunk 1 above)
Comment (jakiejj) [reply src/detext/run_detext.py:191]: This method is deprecated and not used. 
Mainly kept for backward compatibility.

Comment (jakiejj) [reply src/detext/run_detext.py:191]: Misread the comment, but yeah, sure.

File: src/detext/run_detext.py (hunk 2 above)
Comment (jakiejj) [reply src/detext/run_detext.py:156]: `__late_init__` will be called after an argument was created. It's used for post-processing, exactly for the purpose of the refactored code. will add some comments.

**Comment by jakiejj** [comment]
> Thanks for making smart-arg an external dependency for clean repo, and congrats to the package release.
> 
> Apart from the arg parser, there are also a lot changes related to the refactoring. Can you add them in your description as well?

It's already in the `List of changes`, but I can add more details if needed.

**Comment by StarWang** [comment]
LGTM. Can you also add the test information for the two bash tests?

**Comment by jakiejj** [comment]
> LGTM. Can you also add the test information for the two bash tests?

What do you mean?

**Comment by StarWang** [comment]
> > LGTM. Can you also add the test information for the two bash tests?
> 
> What do you mean?

Can you run the two .sh tests under the test/resources folder to make sure the cmd line tests are successful as well?

**Comment by jakiejj** [comment]
> > > LGTM. Can you also add the test information for the two bash tests?
> > 
> > 
> > What do you mean?
> 
> Can you run the two .sh tests under the test/resources folder to make sure the cmd line tests are successful as well?

There is only 1 and it works fine. 
The test_run_detext.py already covered everything. The only additional thing with .sh is converting the command line to a list of str, which is control by the python code.

**Comment by jakiejj** [comment]
> > > > > LGTM. Can you also add the test information for the two bash tests?
> > > > 
> > > > 
> > > > What do you mean?
> > > 
> > > 
> > > Can you run the two .sh tests under the test/resources folder to make sure the cmd line tests are successful as well?
> > 
> > 
> > There is only 1 and it works fine.
> > The test_run_detext.py already covered everything. The only additional thing with .sh is converting the command line to a list of str, which is control by the python code.
> 
> Thanks for running the test.
> 
> As this pr involves arg changes, it's better to run the sh test as well. Although the mechanism behind the cmd call and the test_run_detext.py is the same, args used in the sh file may not be the same as the python tests. Considering that the sh file will be the main entry of running detext for new users, I made the suggestion to be safe.

Nothing in the rb would have this kind of effect though. 
Since this is essential and args might be different, I added it to the unit tests.

File: test/test_run_detext.py [hunk 3]
```diff
@@ -131,3 +131,10 @@ def test_run_detext_multitask(self):
         sys.argv[1:] = args
         main(sys.argv)
         self._cleanUp(output)
+
+
+def test_demo():
```
Comment (StarWang) [inline test/test_run_detext.py:136]: Is there missing indent? This should be included into the class method?

Comment (jakiejj) [reply test/test_run_detext.py:136]: it's a top-level test, as it's not testing the python file directly.

**Pull request merged**